### ISSUE TITLE
PhantomJS is blocking tests to run

### ISSUE THREAD

**Issue by leobalter** [qunitjs/qunit#1048]
```
    Running "qunit:qunit" (qunit) task
    Testing test/index.html
    PhantomJS threw an error:ERROR
    >> 0 [ '' ]
    Warning: PhantomJS exited unexpectedly with exit code null. Use --force to continue.

    Aborted due to warnings.
```

I have no idea what is causing this, some sort of ERROR is happening there where I can't reproduce in a browser test.

I guess it's due to my recent installed Mac OS Sierra.

It works on Ubuntu.

**Comment by trentmwillis** [comment]
@leobalter did this ever resolve? Planning to update my Mac tomorrow, so I'll see if I can reproduce if you're still having issues.

**Comment by leobalter** [comment]
I'm using my home mac for QUnit. I believe we will need to update the grunt Istanbul task by ourselves if want to fix it.

**Comment by mgol** [comment]
I'm seeing the same, also in the Migrate repo.

The problem is that the task is defined in the [grunt-qunit-istanbul](https://github.com/asciidisco/grunt-qunit-istanbul) fork of the upstream [https://github.com/gruntjs/grunt-contrib-qunit](https://github.com/gruntjs/grunt-contrib-qunit) and the fork doesn't seem to be maintained anymore.

**Comment by mgol** [comment]
The [grunt-qunit-istanbul](https://github.com/asciidisco/grunt-qunit-istanbul) fork is also blocking us from upgrading Grunt to version 1.x.

**Comment by mgol** [comment]
...and it still relies on PhantomJS 1 while the upstream repo switched to version 2.

If the Istanbul stuff still can't be done without hacking the upstream `qunit` task we may need to maintain the fork ourselves.

**Comment by trentmwillis** [comment]
I looked into this and it was a bear to figure out what was going on. But I have two PRs now that should enable us to move to Grunt 1.x and run on latest Mac OS:

- https://github.com/asciidisco/grunt-lib-phantomjs-istanbul/pull/11
- https://github.com/asciidisco/grunt-qunit-istanbul/pull/60

**Comment by leobalter** [comment]
This issue is almost solved. Our current problem right now is to figure out a way to have code coverage and not being stuck on PhantomJS' race conditions.

**Comment by platinumazure** [comment]
If we had an lcov reporter, could we use Istanbul?

**Comment by trentmwillis** [comment]
I believe a lot of the complexity in this area could be reduced if we had a separate task to build an instrumented version of QUnit and test against it. The previous tooling tries to do a request-time replacement of the built code which is where the race conditions are coming in.

Once we have CI green, I'll look into a simple approach for this using Istanbul.

**Comment by leobalter** [comment]
> If we had an lcov reporter, could we use Istanbul?

The complex part is getting the instrumented files.

I believe the approach suggested by @trentmwillis is going to be cleaner than what we've had before. Hopefully it will allow us to check coverage on Node too.

**Comment by trentmwillis** [comment]
I worked on this a bit last night. I have code coverage working for Node tests and believe I can get the browser ones working with only a little more infrastructure. We may want to take the outcome of this work and revamp how `grunt-qunit-istanbul` works.

**Comment by trentmwillis** [comment]
Continuing to dig into this. I think I understand the problem space now. I think `grunt-qunit-istanbul` and `grunt-phantomjs-istanbul` were born out of the core `grunt-lib-phantomjs` and `grunt-contrib-qunit` libraries not being easily extensible.

I've opened one PR against `grunt-contrib-qunit` that should allow us to do code coverage within our own repo with relatively minimal setup: https://github.com/gruntjs/grunt-contrib-qunit/issues/135

**Comment by alexweissman** [comment]
@trentmwillis were you able to find a workaround for this?  I've inherited a codebase with a legacy `Gruntfile.js` and testing system, which is generating this error when I try to run the qunit tests.

**Comment by trentmwillis** [comment]
@alexweissman yes, but I don't remember what it was. Your best bet at this point would be to look through the code to see how we have stuff set up.

**Comment by alexweissman** [comment]
Thanks!  I was in chat and someone suggested that PhantomJS might be outdated/obsolete anyway:

>  You might want to try using Karma to run the tests in a real browser (e.g. Chrome or Firefox). grunt-contrib-qunit and PhantomJS are fairly old and might not give a realistic result for end-users.

**Comment by trentmwillis** [comment]
Yes, I agree with that. FWIW, in my projects, I tend to use [Testem](https://github.com/testem/testem) + QUnit these days. Easy to get set up and running.

**Comment by infolock** [comment]
I hope I don't get too much backlash here but it should be something that is noted for any other poor (future) souls who happen to inherit a legacy build that is running into this issue.

If you are (un?)lucky enough to be on a project that may have a dependency on `grunt-contrib-qunit: ~0.7.0` - the way I was able to fix this was by just changing the dependency to `~1.0.0` - which is where the fix for the above was implemented (they stopped support of `0.7.0` and just went straight to `1`).

sorry for this getting re-opened.  i can close it but just wanted to add this note here (for my own sanity as well one day...)
